### ISSUE TITLE
Reactor Skips Over First Level Events Spawned by Reactions. Documentation/Feature Request

### ISSUE THREAD

**Issue by stanvarlamov** [saltstack/salt#36319]
### Description of Issue/Question

If I run an Orchestration from a Web Hook and want to catch and process the Orchestration's return code via the common Reactor system - current design prevents me from doing this directly as the Reactor will ignore "first level" events spawned directly from Reactions (by virtue of 'user': 'Reactor' property set in the event's data). So, it will not process the Orchestration's "ret" event.

Firstly, I could not find this documented - causes a major headache designing, testing the Orchestration directly via salt-run, then finding out that this doesn't work via the indirect launch and spinning wheels trying to figure out what is going on and how to mitigate it.

Second, I believe that the inconsistency in handling Events breaks the modular design approach - now I have to accommodate for the fact that some events will or will not be handled by the Reactor depending on where the process was invoked from.

I understand the current design was put in place to control potential recursive event firing situations and cannot be changed for backward compatibility, but I'm wondering if it's possible to add a flag into the Reaction SLS to force processing of the first level event (and document that too)?

As a workaround, I have to either hook into the event queue directly or try putting logic into a "secondary" event (which I don't think I really can, because I can't reliably control the process will get to that point if something else in the Orchestration crashes and causes it to quit) , but I believe this is quite fundamental to just leave it as-is.
### Steps to Reproduce Issue

Put an Orchestration into a Reaction and watch in debug log its /salt/run/ event pair ignored by the Reactor
### Versions Report

`Salt Version:
           Salt: 2016.3.0-n/a-440116c

Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.2.2
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.1
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-92-generic
         system: Linux
        version: Ubuntu 14.04 trusty
`

**Comment by Ch3LL** [comment]
regardless I think this is a documentation bug. To answer your questions about whether we could add a flag to ensure first level events are processed i'm gonna have to refer to @cachedout on this one. Do you think this would be something we could implement?

**Comment by sv852** [comment]
Thank you for looking into this. #30604 is the latest I found where @jacksontj explains "this feature was added to avoid loops. ... have a reactor that runs on every event you will infinitely recurse. ... we could look into some other mechanism to not fire events from the reactor."  
I have to respectfully disagree with the presumed necessity to block firing events from the reactor (what I refer to as "first level" events in my comment). To my argument, "for-loops" are not banned from languages just because they can become infinite under poor coding techniques.

By adding a simple check into the Reaction SLS I can easily avoid recursivity, e.g.:
- a web hook triggers a Reaction that fires runner.state.orchestrate directly from its SLS
- another reactor is set to watch for salt/run/*/ret events, and its Reaction launches a different type of runner (not runner.state.orchestrate)
- this code is commented out in salt.utils.reactor.py:
`# skip all events fired by ourselves
            if data['data'].get('user') == self.wrap.event_user:
                continue`
- the system keeps on executing events and reactions indefinitely till the service is killed - confirms the recursivity problem when the above safety code is removed
- however, if I put this into salt/run/*/ret event's Reaction to bypass any processing unless the event was fired for runner.state.orchestrate
`{% if data.fun == 'runner.state.orchestrate' %}`
the process exits gracefully, after processing the final condition for the runner.state.orchestrate event I'm interested in

Back to the feature request, if I can control whether or not the Reaction-fired event is put back into the Reactor's loop - I can carefully build Reactions to interrogate all events I'm interested in, or exit otherwise.

**Comment by Ch3LL** [comment]
Thanks for all the great investigation work on this one. As stated before I want to get @cachedout 's opinion on this feature request and the above information will help to determine whether we approve. Thanks

**Comment by jacksontj** [comment]
@sv852 I actually completely agree with you. The reason for the prior patch was because there was no mechanism to *stop* the recursion, and since it was "broken" before-- it was fine to do so. Keep in mind though, there are a decent amount of little problems here to deal with. The main one being that the events fired are down in other layers of the code-- so its hard to control what they do.

That all being said, I'm all for creating better scoping in the reactor-- and it would be nice to allow for some actions based on reactor actions. The ask for "control whether or not the Reaction-fired event is put back into the Reactor's loop" is actually a fairly difficult ask-- PRs are welcome, I'd be glad to talk over ideas/options as well :)

**TLDR; i agree-- we don't have to _disallow_ recursion, its just complicated to support in the current reactor event setup**

**Comment by sv852** [comment]
@jacksontj thank you for clarifying. An lD-level flag in Reaction SLS to explicitly mark steps for the Reactor system to keep in the pool - can be complex, as you're pointing out, and excessive. I'm wondering if a Master Configuration level flag would do? By default, keep the current approach of bypassing "self-generated" events in the Reactor, but if the flag is set - react to all events. I think that would just be another piece in the "if" above the current "continue" that controls the skip.

Elaborating further, the web hook functionality that brought this issue to my attention sounds very promising as a mechanism to employ Salt in, e.g., cloud-generation process triggered by an outside system that in return gets notified via web hook-style communications, as opposite to API request-reply, not suited well for long-running processes. However, I see a number of side issues that have to be resolved before this becomes production-strength, and I may have to ultimately opt for a different approach.

In summary, in closure of this thread I suggest:
- update the documentation to point out that immediate Reaction-generated events are bypassed by the Reactor (as already marked by @Ch3LL here) 
- design and document an optional Master Configuration flag to suppress the above behavior if desired so by the developer, who will bear consequences and code the Reactions to control infinite processes as needed
